### PR TITLE
fix: do not transform ClassPrivateMethods in estree

### LINKED ISSUE #11972: visitor keys for `ClassPrivateMethod` type do not match the structure of `ClassPrivateMethod` nodes
Originally reported at https://github.com/eslint/eslint/issues/13575.

## Bug Report

- [x] I would like to work on a fix!

**Current behavior**

Using `@babel/eslint-parser` to parse some code with a private method declaration, `eslint` crashes:

```js
Oops! Something went wrong! :(

ESLint: 7.7.0

TypeError: Cannot read property 'type' of undefined
Occurred while linting /home/piranna/Trabajo/Veedeo.me/kurento-utils-js/src/WebRtcPeerCore.js:30
    at collectUnusedVariables (/home/piranna/Trabajo/Veedeo.me/kurento-utils-js/node_modules/eslint/lib/rules/no-unused-vars.js:569:50)
    at collectUnusedVariables (/home/piranna/Trabajo/Veedeo.me/kurento-utils-js/node_modules/eslint/lib/rules/no-unused-vars.js:603:17)
    at collectUnusedVariables (/home/piranna/Trabajo/Veedeo.me/kurento-utils-js/node_modules/eslint/lib/rules/no-unused-vars.js:603:17)
    at collectUnusedVariables (/home/piranna/Trabajo/Veedeo.me/kurento-utils-js/node_modules/eslint/lib/rules/no-unused-vars.js:603:17)
    at Program:exit (/home/piranna/Trabajo/Veedeo.me/kurento-utils-js/node_modules/eslint/lib/rules/no-unused-vars.js:615:36)
    at /home/piranna/Trabajo/Veedeo.me/kurento-utils-js/node_modules/eslint/lib/linter/safe-emitter.js:45:58
    at Array.forEach (<anonymous>)
    at Object.emit (/home/piranna/Trabajo/Veedeo.me/kurento-utils-js/node_modules/eslint/lib/linter/safe-emitter.js:45:38)
    at NodeEventGenerator.applySelector (/home/piranna/Trabajo/Veedeo.me/kurento-utils-js/node_modules/eslint/lib/linter/node-event-generator.js:254:26)
    at NodeEventGenerator.applySelectors (/home/piranna/Trabajo/Veedeo.me/kurento-utils-js/node_modules/eslint/lib/linter/node-event-generator.js:283:22)
```

**Expected behavior**

Parsing should be done flawlessly.

**Babel Configuration**
- Filename: `babel.config.js`

```js
module.exports = {
  presets: ['@babel/preset-env'],
  plugins: [
    "@babel/plugin-proposal-class-properties",
    "@babel/plugin-proposal-private-methods"
  ]
}
```

**Environment**

```
  System:
    OS: Linux 5.4 Ubuntu 20.04 LTS (Focal Fossa)
  Binaries:
    Node: 14.4.0 - /usr/bin/node
    npm: 6.14.5 - /usr/bin/npm
```

- Babel version(s): 7.10.5
- Monorepo: no
- How you are using Babel: `cli`

**Possible Solution**

Add `parent` field to `ClassPrivateMethod` nodes

**Additional context**

Originally reported (and with more info) at https://github.com/eslint/eslint/issues/13575

### PR DESCRIPTION
**Pull request by JLHwung**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #11972 
| Patch: Bug Fix?          | Yes
| Tests Added + Pass?      | Yes
| License                  | MIT

Use `MethodDefinition` as the visitorKeys of `ClassPrivateMethods`. This should be considered as a ad-hoc fixes before the class features support is landed.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/11973"></a>

### CODE REVIEW

**Comment by mdjermanovic** [comment]
This change might break ESLint's code path analysis, and probably some other places where it is assumed that only`FunctionDeclaration`, `FunctionExpression`, and `ArrowFunctionExpression` can represent a function.

```js
/* eslint no-unreachable: error */

class C {
  #a() {
    return;
  }
  #b() {} // no-unreachable false positive
}
```

File: eslint/babel-eslint-parser/src/convert/convertAST.js [hunk 1]
```diff
@@ -81,30 +81,30 @@ function convertNodes(ast, code) {
...
-  const oldExportAllDeclarationKeys = t.VISITOR_KEYS.ExportAllDeclaration;
+  const oldVisitorKeys = new Map();
 
   try {
-    // Monkey patch visitor keys in order to be able to traverse the estree nodes
-    t.VISITOR_KEYS.ChainExpression = VISITOR_KEYS.ChainExpression;
-    t.VISITOR_KEYS.ImportExpression = VISITOR_KEYS.ImportExpression;
-    t.VISITOR_KEYS.Property = VISITOR_KEYS.Property;
-    t.VISITOR_KEYS.MethodDefinition = VISITOR_KEYS.MethodDefinition;
+    for (const [type, visitorKey] of Object.entries(conflictTypes)) {
```
Comment (JLHwung) [inline eslint/babel-eslint-parser/src/convert/convertAST.js:87]: Also refactored a bit to remove boilerplates.

Comment (kaicataldo) [reply eslint/babel-eslint-parser/src/convert/convertAST.js:87]: Nice refactor!

**Comment by piranna** [comment]
Thanks for merging it :-) What's the stimated time to publish the next 7.11.4 version with this fix?

**Pull request merged**